src/state-manager/TransactionConsensus.ts

```diff
@@ -1024,6 +1024,239 @@ class TransactionConsenus {
+            Comms.sendGossip(
+              'poqo-receipt-gossip',
+              payload.receipt,
+              null,
+              null,
+              queueEntry.transactionGroup,
+              false,
+              4,
+              payload.finalState.txid,
+              '',
+              true
+            )
+            queueEntry.hasSentFinalReceipt = true
+          }
+        } finally {
+          profilerInstance.scopedProfileSectionEnd('poqo-data-and-receipt')
+        }
+      }
+    )
+
+    Comms.registerInternal(
+      'poqo-send-receipt',
+      (
+        payload: AppliedReceipt2,
+        _respond: unknown,
+        _sender: unknown,
+        _tracker: string,
+        msgSize: number
+      ) => {
+        profilerInstance.scopedProfileSectionStart('poqo-send-receipt', false, msgSize)
+        try{
+          const queueEntry = this.stateManager.transactionQueue.getQueueEntrySafe(payload.txid)
+          if (queueEntry == null) {
+            /* prettier-ignore */ nestedCountersInstance.countEvent('poqo', 'poqo-send-receipt: no queue entry found')
+            return
+          }
+
+          if (queueEntry.poqoReceipt) {
+            // We've already handled this
+            return
+          }
+
+          if (logFlags.verbose) this.mainLogger.debug(`POQo: Received receipt from aggregator for ${queueEntry.logID} starting CT2 for data & receipt`)
+          const receivedReceipt = payload as AppliedReceipt2
+          queueEntry.poqoReceipt = receivedReceipt
+          queueEntry.appliedReceipt2 = receivedReceipt
+          queueEntry.recievedAppliedReceipt2 = receivedReceipt
+          queueEntry.hasSentFinalReceipt = true
+          Comms.sendGossip(
+            'poqo-receipt-gossip',
+            payload,
+            null,
+            null,
+            queueEntry.transactionGroup,
+            false,
+            4,
+            payload.txid,
+            '',
+            true
+          )
+          this.stateManager.transactionQueue.factTellCorrespondingNodesFinalData(queueEntry)
+        } finally {
+          profilerInstance.scopedProfileSectionEnd('poqo-send-receipt')
+        }
+      }
+    )
+
+    Comms.registerInternal(
+      'poqo-send-vote',
+      async (
+        payload: AppliedVoteHash,
+        _respond: unknown,
+        _sender: unknown,
+        _tracker: string,
+        msgSize: number
+      ) => {
+        profilerInstance.scopedProfileSectionStart('poqo-send-vote', false, msgSize)
+        try{
+          const queueEntry = this.stateManager.transactionQueue.getQueueEntrySafe(payload.txid)
+          if (queueEntry == null) {
+            /* prettier-ignore */ nestedCountersInstance.countEvent('poqo', 'poqo-send-vote: no queue entry found')
+            return
+          }
+          const collectedVoteHash = payload as AppliedVoteHash
+          // We can reuse the same function for POQo
+          this.tryAppendVoteHash(queueEntry, collectedVoteHash)
```

Comment:
We should also check that sign of the voteHash is from the executionGroupNodes.
A malicious executionGroup can create bad voteHash and sign by using different keys, we should check if all these sign owner are in the exectionGroupNodes and reject them if they are not there.